src/main/java/com/aws/greengrass/status/FleetStatusService.java

```diff
@@ -298,36 +304,71 @@ public void schedulePeriodicFleetStatusDataUpdate(boolean isDuringConnectionResu
+            // be in-progress and other component's fleet config arn may not have updated
+            Set<GreengrassService> erroredcomponet = new HashSet<>();
+            erroredcomponet.add(greengrassService);
+            if (isDeploymentInProgress.get()) {
+                uploadFleetStatusServiceData(erroredcomponet, overAllStatus.get(), null,
+                        Trigger.COMPONENT_STATUS_CHANGE);
+                return;
+            }
+            // Report status of other components, in case recovery duration takes too long and other components need
+            // to wait until all components to reach terminal state to update status
+            synchronized (updatedGreengrassServiceSet) {
+                uploadFleetStatusServiceData(updatedGreengrassServiceSet, overAllStatus.get(), null,
+                        Trigger.COMPONENT_STATUS_CHANGE);
+            }
             return;
         }
 
-        synchronized (updatedGreengrassServiceSet) {
-            updatedGreengrassServiceSet.add(greengrassService);
+        if (!isDeploymentInProgress.get()) {
+            // if there is no ongoing deployment and we encounter a BROKEN component,
+            // update the fleet status as UNHEALTHY.
+            if (newState.equals(State.BROKEN)) {
+                synchronized (updatedGreengrassServiceSet) {
+                    uploadFleetStatusServiceData(updatedGreengrassServiceSet, OverallStatus.UNHEALTHY,
+                            null, Trigger.COMPONENT_STATUS_CHANGE);
+                }
+            }
+            // Check and update component status when a component reaches to terminal state
+            if (greengrassService.reachedDesiredState()) {
+                handleServiceReachedTerminalState();
+            }
         }
+    }
 
-        // if there is no ongoing deployment and we encounter a BROKEN component, update the fleet status as UNHEALTHY.
-        if (!isDeploymentInProgress.get() && newState.equals(State.BROKEN)) {
-            synchronized (updatedGreengrassServiceSet) {
-                uploadFleetStatusServiceData(updatedGreengrassServiceSet, OverallStatus.UNHEALTHY,
-                        null, Trigger.COMPONENT_STATUS_CHANGE);
+    private void handleServiceReachedTerminalState() {
+        //1. check all other components reached terminal state
+        boolean allReachedTerminalState = kernelLifecycle.allServicesInTerminalState();
+        //2. If all other components are in terminal state, update status for all component
+        // Skip updates when kernel shutdown since service ends with FINISHED for stopping
+        synchronized (updatedGreengrassServiceSet) {
+            if (!updatedGreengrassServiceSet.isEmpty() && allReachedTerminalState && !isDeploymentInProgress.get()
+                    && !kernelLifecycle.getIsShutdownInitiated().get()) {
+                Instant now = Instant.now();
```

Comment:
nit: seems like this is a duplicate of the code at 312. maybe extract this into a method?

Reply:
Extracting them into a method creates bugs :(

Reply:
With refactoring, the `updatedGreengrassServiceSet` will not always contain the correct values, I observed that sometimes deployment fss message doesn't contain any component details, which means `updatedGreengrassServiceSet` already got cleared. There will be thread timing issue

Reply:
I think we need to address this duplicate code. I don't understand why updatedGreengrassServiceSet can cause race condition. You can extract these codes in a way that does not depend on this set